tiny_dnn/core/framework/tensor.h

```diff
@@ -61,45 +63,173 @@
+     * @param shape array containing N integers, sizes of dimensions
+     * @return
+     */
+    explicit TensorStorage(std::initializer_list<size_t> const &shape) {
+        std::vector<size_t> tmp(shape.size());
+        std::copy(shape.begin(), shape.end(), tmp.begin());
+        resize(tmp);
+    }
+
+    void sync() {
+        if (data_is_on_host_)
+            toDevice();
+        else
+            fromDevice();
+    }
+
+    typename std::vector<U, aligned_allocator < U, 64>>::
+    iterator host_data(size_t offset) {
+        return host_data_.begin() + offset;
+    }
+
+    typename std::vector<U, aligned_allocator < U, 64>>::
+    const_iterator host_data(size_t offset) const {
+        return host_data_.begin() + offset;
+    }
+
+    void resize(const std::vector<size_t> &sz) {
+        host_data_->resize(std::accumulate(std::begin(sz),
+                                           std::end(sz),
+                                           size_t(1),
+                                           std::multiplies<size_t>()), U(0));
+        //TODO(Randl): device data
+    }
+
+ private:
+
+    void toDevice() const {
+        if (data_is_on_host_ && data_dirty_) {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+            CLCudaAPI::Queue queue = device_->queue();
+            if (device_data_ && device_data_->GetSize() >= host_data_->size()) {
+                device_data_->Write(queue,
+                                    host_data.size(),
+                                    host_data_->data(),
+                                    0);
+            } else {
+                CLCudaAPI::Context ctx = device_->context();
+                device_data_ = make_unique<CLCudaAPI::Buffer < U> > (
+                    ctx, queue, host_data_->begin(), host_data_->end());
+            }
+#endif
+
+            data_is_on_host_ = false;
+            data_dirty_ = false;
+        }
+    }
+
+    void fromDevice() const {
+        if (!data_is_on_host_ && data_dirty_) {
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+            assert(device_);
+            assert(device_data_);
+            device_data_->Read(device_->queue(),
+                               host_data_->size(),
+                               // using const_cast<> to avoid making host_data_
+                               // entirely mutable
+                               const_cast<U *>(host_data_->data()));
+#endif
+            data_is_on_host_ = true;
+            data_dirty_ = false;
+        }
+    }
+
+    std::vector<U, aligned_allocator<U, 64>> host_data_;
+
+#if defined(USE_OPENCL) || defined(USE_CUDA)
+    /* Pointer to the Tensor data in the device */
+    std::unique_ptr<CLCudaAPI::Buffer<U>> device_data_;
+#endif
+    mutable bool data_is_on_host_; // is current data is on host?
+    mutable bool data_dirty_;      // have current data might been modified?
+    mutable size_t dirty_from, dirty_to; // range of indexes that can be modified
```

Comment:
I didn't know that comment lines can be longer than 80 characters. So you don't need to fix this one..
https://google.github.io/styleguide/cppguide.html#Line_Length

Hmmm I guess that means we can hide wide codes in comments and apply preprocessing tool to expand it before compilers strip out the comments.